File: src/main/java/com/violeth/blockjs/blockjs/jsinterface/mcinterface/Player.java
package com.violeth.blockjs.blockjs.jsinterface.mcinterface;

import com.eclipsesource.v8.V8;
import com.eclipsesource.v8.V8Array;
import com.eclipsesource.v8.V8Function;
import com.violeth.blockjs.blockjs.BlockJS;
import org.bukkit.Bukkit;
import org.bukkit.GameMode;

import java.util.Collection;
import java.util.UUID;

public class Player {
    V8 v8;
    public Player(V8 v8) {
        this.v8 = v8;
    }
    public void setGameMode(String playerUUID, int gameMode) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));
        var gameModeEnumValue = GameMode.getByValue(gameMode);

        if (player != null && gameModeEnumValue != null) {
            player.setGameMode(gameModeEnumValue);
        }
    }
    public void setWalkSpeed(String playerUUID, float speed) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.setWalkSpeed(speed);
        }
    }
    public void setFlySpeed(String playerUUID, float speed) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.setFlySpeed(speed);
        }
    }
    public void setFlying(String playerUUID, boolean state) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.setFlying(state);
        }
    }
    public void setSneaking(String playerUUID, boolean state) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.setSneaking(state);
        }
    }
    public void setSprinting(String playerUUID, boolean state) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.setSprinting(state);
        }
    }
    public void setHealth(String playerUUID, double health) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.setHealth(health);
        }
    }
    public void setExhaustion(String playerUUID, float exhaustion) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.setExhaustion(exhaustion);
        }
    }
    public void setPosition(String playerUUID, double x, double y, double z) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            var location = player.getLocation();

            location.setX(x);
            location.setY(y);
            location.setZ(z);
        }
    }
    public void setVelocity(String playerUUID, double x, double y, double z) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            var velocity = player.getVelocity();

            velocity.setX(x);
            velocity.setY(y);
            velocity.setZ(z);
        }
    }
    public String getOnlinePlayerUUIDByName(String name) {
        var player = Bukkit.getPlayer(name);

        if(player != null) {
            return player.getUniqueId().toString();
        }

        return null;
    }

    public String getOfflinePlayerUUIDByName(String name) {
        var player = Bukkit.getOfflinePlayer(name);

        if(player != null) {
            return player.getUniqueId().toString();
        }

        return null;
    }

    public V8Array getOnlinePlayersUUIDs() {
        var players = org.bukkit.Bukkit.getOnlinePlayers();
        var uuids = new V8Array(v8);

        for(var playersIt = players.iterator(); playersIt.hasNext();) {
            uuids.push(playersIt.next().getUniqueId().toString());
        }

        return uuids;
    }

    public String getPlayerNameByUUID(String playerUUID) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if(player != null) {
            return player.getName();
        }

        return null;
    }

    public void doDamage(String playerUUID, double damage) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.damage(damage);
        }
    }

    public void heal(String playerUUID, double heal) {
        var player = Bukkit.getPlayer(UUID.fromString(playerUUID));

        if (player != null) {
            player.setHealth(player.getHealth() + heal);
        }
    }
}

File: plugin/src/main/kotlin/io/github/blockjs/plugin/js/runtime/api/game/Player.kt
package io.github.blockjs.plugin.js.runtime.api.game

import com.caoccao.javet.annotations.V8Function
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.entity.Player
import java.util.UUID

@Suppress("unused")
object Player {
    @V8Function
    fun setGameMode(playerUUID: String?, gameMode: Int) {
        val player  = Bukkit.getPlayer(UUID.fromString(playerUUID))
        val gameModeEnumValue = GameMode.getByValue(gameMode)

        if (player != null && gameModeEnumValue != null) {
            player.gameMode = gameModeEnumValue
        }
    }

    @V8Function
    fun setWalkSpeed(playerUUID: String, speed: Float) {
        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))

        if (player != null) {
            player.walkSpeed = speed
        }
    }

    @V8Function
    fun setFlySpeed(playerUUID: String, speed: Float) {
        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))

        if (player != null) {
            player.flySpeed = speed
        }
    }

    @V8Function
    fun setFlying(playerUUID: String?, state: Boolean) {
        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))

        if (player != null) {
            player.isFlying = state
        }
    }

    @V8Function
    fun setSneaking(playerUUID: String, state: Boolean) {
        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))

        if (player != null) {
            player.isSneaking = state
        }
    }

    @V8Function
    fun setSprinting(playerUUID: String, state: Boolean) {
        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))

        if (player != null) {
            player.isSprinting = state
        }
    }

    @V8Function
    fun setHealth(playerUUID: String, health: Double) {
        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))

        if (player != null) {
            player.health = health
        }
    }

    @V8Function
    fun setExhaustion(playerUUID: String, exhaustion: Float) {

        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))

        if (player != null) {
            player.exhaustion = exhaustion
        }
    }

    @V8Function
    fun setPosition(playerUUID: String, x: Double, y: Double, z: Double) {
        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))
        if (player != null) {
            val location = player.location

            location.x = x
            location.y = y
            location.z = z
        }
    }

    @V8Function
    fun setVelocity(playerUUID: String, x: Double, y: Double, z: Double) {
        val player = Bukkit.getPlayer(UUID.fromString(playerUUID))

        if (player != null) {
            val velocity = player.velocity

            velocity.x = x
            velocity.y = y
            velocity.z = z
        }
    }

    @V8Function
    fun doDamage(uuid: String, damage: Double) {
        val player = Bukkit.getPlayer(UUID.fromString(uuid))

        if (player != null) {
            player.damage(damage)
        }
    }

    @V8Function
    fun heal(uuid: String, heal: Double) {
        val player = Bukkit.getPlayer(UUID.fromString(uuid))

        if (player != null) {
            player.health += heal
        }
    }
}
